crates/core/src/surfnet/mod.rs

```diff
@@ -0,0 +1,696 @@
+    ///
+    /// # Parameters
+    ///
+    /// - `pubkey`: The public key of the account to retrieve.
+    /// - `strategy`: The strategy to use for fetching the account (`LocalOrDefault`, `ConnectionOrDefault`, etc.).
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing an optional account, which may be `None` if the account was not found.
+    pub async fn get_account(
+        &self,
+        pubkey: &Pubkey,
+        strategy: GetAccountStrategy,
+    ) -> Result<Option<Account>, Box<dyn std::error::Error>> {
+        let (result, factory) = match strategy {
+            GetAccountStrategy::LocalOrDefault(factory) => {
+                (self.inner.get_account(pubkey), factory)
+            }
+            GetAccountStrategy::ConnectionOrDefault(factory) => {
+                let client = self.expected_rpc_client();
+                let entry = client.get_account(pubkey).await?;
+                (Some(entry), factory)
+            }
+            GetAccountStrategy::LocalThenConnectionOrDefault(factory) => {
+                match self.inner.get_account(pubkey) {
+                    Some(entry) => (Some(entry), factory),
+                    None => {
+                        let client = self.expected_rpc_client();
+                        let entry = client.get_account(pubkey).await?;
+                        (Some(entry), factory)
+                    }
+                }
+            }
+        };
+        let account = match (result, factory) {
+            (None, Some(factory)) => Some(factory(self)),
+            (None, None) => None,
+            (Some(account), _) => Some(account),
+        };
+        Ok(account)
+    }
+
+    /// Retrieves a mutable account for the specified public key based on the given strategy.
+    ///
+    /// This function works similarly to `get_account`, but will mutate the underlying state with the Account.
+    /// Note: if the requested account is executable, the data account is also retrieved and stored.
+    ///
+    /// # Parameters
+    ///
+    /// - `pubkey`: The public key of the account to retrieve.
+    /// - `strategy`: The strategy to use for fetching the account (`LocalOrDefault`, `ConnectionOrDefault`, etc.).
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing an optional mutable account.
+    pub async fn get_account_mut(
+        &mut self,
+        pubkey: &Pubkey,
+        strategy: GetAccountStrategy,
+    ) -> Result<Option<Account>, Box<dyn std::error::Error>> {
+        let (result, factory) = match strategy {
+            GetAccountStrategy::LocalOrDefault(factory) => {
+                (self.inner.get_account(pubkey), factory)
+            }
+            GetAccountStrategy::ConnectionOrDefault(factory) => {
+                let client = self.expected_rpc_client();
+                let account = client.get_account(&pubkey).await?;
+                let _ = self.inner.set_account(pubkey.clone(), account.clone());
```

Comment:
Should we also be get/setting the program data address here too if executable?

Reply:
Addressed with latest